### PR TITLE
[TTS][DE] Multi-speaker fastpitch model training recipe on HUI-Audio-Corpus-German

### PR DESCRIPTION
**Pull request by XuesongYang**
# What does this PR do ?
* created a dataset preparation recipe for HUI-ACG.
    * re-write `get_data.py` to support multiprocessing and decouple text normalization. 
* added a script to convert graphemes into phonemes.
* created a model config for multi-speaker Fastpitch model training.
 

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```bash
# Download the dataset, normalize texts, create train/val/test splits of JSON manifests.
$ python scripts/dataset_processing/tts/hui_acg/get_data.py \
        --data-root /home/xueyang/datasets \
        --normalize-text
$ ls /home/xueyang/datasets/HUI-Audio-Corpus-German-clean/*_text_normed.json
test_manifest_text_normed.json  train_manifest_text_normed.json  val_manifest_text_normed.json

# run phonemizer to convert graphemes into phonemes, resulting in new files with name like train_manifest_text_normed_phonemes.json
$ python scripts/dataset_processing/tts/hui_acg/phonemizer.py \
        --json-manifests /home/xueyang/datasets/HUI-Audio-Corpus-German-clean/train_manifest_text_normed.json \
                                     /home/xueyang/datasets/HUI-Audio-Corpus-German-clean/val_manifest_text_normed.json \
                                     /home/xueyang/datasets/HUI-Audio-Corpus-German-clean/test_manifest_text_normed.json \
        --preserve-punctuation
$ ls /home/xueyang/datasets/HUI-Audio-Corpus-German-clean/*_phonemes.json
test_manifest_text_normed_phonemes.json  train_manifest_text_normed_phonemes.json  val_manifest_text_normed_phonemes.json

$ python scripts/dataset_processing/tts/extract_sup_data.py \
        --config-path hui_acg/ds_conf \
        --config-name ds_for_fastpitch_align.yaml \
        manifest_filepath=/home/xueyang/datasets/HUI-Audio-Corpus-German-clean/Friedrich_Clean/train_manifest_text_normed_phonemes.json \
        sup_data_path=/home/xueyang/datasets/HUI-Audio-Corpus-German-clean/sup_data_phonemes

# you would see the information in the terminal
-- Set up ENVs...
a867b5ef4cd08fe0db216937b2d3e767fe71b3e7
-- Started ...
[NeMo I 2022-06-17 15:15:57 tokenize_and_classify:81] Creating ClassifyFst grammars. This might take some time...
[NeMo I 2022-06-17 15:16:15 data:186] Loading dataset from /data/HUI-Audio-Corpus-German-clean/train_manifest_text_normed_phonemes.json.
[NeMo I 2022-06-17 15:16:22 data:221] Loaded dataset with 173622 files.
[NeMo I 2022-06-17 15:16:22 data:223] Dataset contains 403.30 hours.
[NeMo I 2022-06-17 15:16:22 data:321] Pruned 0 files. Final dataset contains 173622 files
[NeMo I 2022-06-17 15:16:22 data:323] Pruned 0.00 hours. Final dataset contains 403.30 hours.
Processing /home/xueyang/datasets/HUI-Audio-Corpus-German-clean/train_manifest_text_normed_phonemes.json:
PITCH_MEAN=155.3916473388672, PITCH_STD=102.26656341552734
PITCH_MIN=65.4063949584961, PITCH_MAX=2093.004638671875

# train FastPitch model. copy & paste PITCH_MEAN, PITCH_STD, PITCH_MIN, and PITCH_MAX to the params.
$ python examples/tts/fastpitch.py \
        --config-path conf/de \
        --config-name fastpitch_align_44100 \
        model.train_ds.dataloader_params.num_workers=16 \
        model.validation_ds.dataloader_params.num_workers=16 \
        model.train_ds.dataloader_params.batch_size=24 \
        model.validation_ds.dataloader_params.batch_size=24 \
        train_dataset=/home/xueyang/datasets/HUI-Audio-Corpus-German-clean/train_manifest_text_normed_phonemes.json \
        validation_datasets=/home/xueyang/datasets/HUI-Audio-Corpus-German-clean/val_manifest_text_normed_phonemes.json \
        sup_data_path=/home/xueyang/datasets/HUI-Audio-Corpus-German-clean/sup_data_phonemes \
        pitch_mean=155.3916473388672 \
        pitch_std=102.26656341552734 \
        pitch_fmin=65.4063949584961 \
        pitch_fmax=2093.004638671875 \
        trainer.devices=2 \
        trainer.num_nodes=1 \
        trainer.accumulate_grad_batches=2 \
        trainer.log_every_n_steps=10 \
        exp_manager.exp_dir=/home/xueyang/experiments/GermanTTS/multi_spk_tts_de/fastpitch-train/results \
        exp_manager.resume_if_exists=True \
        exp_manager.resume_ignore_no_checkpoint=True \
        +exp_manager.create_wandb_logger=True \
        +exp_manager.wandb_logger_kwargs.name=fastpitch-train \
        +exp_manager.wandb_logger_kwargs.project=multi_spk_tts_de 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: scripts/dataset_processing/tts/hui_acg/get_data.py [hunk 1]
```diff
@@ -19,156 +19,296 @@
...
+        for line in tqdm(foverview):
             file_stem, duration, *_, text = line.strip().split("|")
             duration = float(duration)
 
-            # file_stem -> dir_name (e.g. maerchen_01_f000051 -> maerchen, ber_psychoanalyse_01_f000046 -> ber_psychoanalyse)
+            # file_stem -> dir_name (e.g. maerchen_01_f000051 -> maerchen)
             dir_name = "_".join(file_stem.split("_")[:-2])
             audio_path = dataset_path / dir_name / "wavs" / f"{file_stem}.wav"
 
             if min_duration <= duration <= max_duration:
```
Comment (subhankar-ghosh) [inline scripts/dataset_processing/tts/hui_acg/get_data.py:145]: This would drop some datapoints from the original dataset, do we want to do this or give the user the option to do so in the get_data.py file?

File: scripts/dataset_processing/tts/hui_acg/phonemizer_local.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    )
+    args = parser.parse_args()
+    return args
+
+
+def main():
+    args = get_args()
+    input_manifest_filepaths = args.json_manifests
+    preserve_punctuation = args.preserve_punctuation
+    # TODO (xueyang): consider with_stress=True and add stress symbols into charset of tokenizer.
```
Comment (subhankar-ghosh) [inline scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:49]: Remove all TODO comments.

File: scripts/dataset_processing/tts/hui_acg/phonemizer_local.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+"""
+The original JSON splits generated from `get_data.py` only contain graphemes inputs. We recommend adding phonemes as
```
Comment (subhankar-ghosh) [inline scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:2]: Also this will add phonemes to the manifests right? Are we sure we want to do that? And is this consistent with how we use or add phonemes in NeMo?

Comment (redoctopus) [reply scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:2]: It seems worth adding as an option, in addition to being able to do phonemizing on the fly.

I recall @yzhang123 and @ekmb have added some support for a specific format it should be in for the normalizer to pass it through untouched, could one of you confirm?

File: scripts/dataset_processing/tts/hui_acg/ds_conf/ds_for_fastpitch_align.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+name: "ds_for_fastpitch_align"
+
+manifest_filepath: "train_manifest_text_normed_phonemes.json"
+sup_data_path: "sup_data_phonemes"
```
[the remark this answers is not in the document]
Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/ds_conf/ds_for_fastpitch_align.yaml:4]: agreed. let's make it `???` for `manifest_filepath` and `sup_data_path`.

File: scripts/dataset_processing/tts/hui_acg/get_data.py (hunk 1 above)
Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/get_data.py:145]: Yes. That's designed on purpose. We provided min and max duration arguments. Please check Line 60 and 61.

File: scripts/dataset_processing/tts/hui_acg/get_data.py [hunk 5]
```diff
@@ -19,156 +19,296 @@
...
+    num_speakers = 0
+    for child in stats_path_root.iterdir():
+        if child.is_dir():
+            speaker = child.name
+            num_speakers += 1
+            speaker_stats_root = stats_path_root / speaker
+            speaker_data_path = dataset_root / speaker
+
+            logging.info(f"Processing Speaker: {speaker}")
+            results = __process_data(
```
[the remark this answers is not in the document]
Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/get_data.py:262]: `joblib.Parallel` actually can use `multiprocessing.pool` backend. It is a higher-level wrapper. We used this in NeMo.

File: scripts/dataset_processing/tts/hui_acg/phonemizer_local.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+"""
```
[the remark this answers is not in the document]
Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:1]: Good point. Will rename it in my next revision.

File: scripts/dataset_processing/tts/hui_acg/phonemizer_local.py (hunk 3 above)
Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:2]: @subhankar-ghosh yes. we will add new entry in the JSON. `{..."normalized_text": xyz, "is_phoneme": 1}`. We decided to use a mixer of grapheme/phoneme for German as suggested because German is kinda a phonetic language.

File: scripts/dataset_processing/tts/hui_acg/phonemizer_local.py (hunk 2 above)
Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:49]: will do.

Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:49]: a second thought is that mixing with phonemes is still experimental so that it would be helpful to leave the `TODO` here for the revision in next iterations.

File: scripts/dataset_processing/tts/hui_acg/phonemizer_local.py (hunk 3 above)
Comment (XuesongYang) [reply scripts/dataset_processing/tts/hui_acg/phonemizer_local.py:2]: The tokenizer using phonemes for German is different from what English used.
